File: modules/340-extended-monitoring/images/extended-monitoring-exporter/src/extended-monitoring.py
#!/usr/bin/env python3

# Copyright 2021 Flant JSC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from concurrent.futures.thread import ThreadPoolExecutor
from datetime import datetime, timedelta
from itertools import chain
from threading import Thread
from time import sleep
import logging

import kubernetes
import copy
import sys

from abc import ABC, abstractmethod
from http.server import BaseHTTPRequestHandler, HTTPServer
from socketserver import ThreadingMixIn

kubernetes.config.load_incluster_config()

logging.basicConfig(format='[%(asctime)s] - %(message)s', level=logging.INFO)

DEPRECATED_EXTENDED_MONITORING_ANNOTATION_THRESHOLD_PREFIX = "threshold.extended-monitoring.flant.com/"
DEPRECATED_EXTENDED_MONITORING_ENABLED_ANNOTATION = "extended-monitoring.flant.com/enabled"
EXTENDED_MONITORING_LABEL_THRESHOLD_PREFIX = "threshold.extended-monitoring.deckhouse.io/"
EXTENDED_MONITORING_ENABLED_LABEL = "extended-monitoring.deckhouse.io/enabled"

DEFAULT_SERVER_ADDRESS = '0.0.0.0'
DEFAULT_PORT = 8080


class ThreadingHTTPServer(ThreadingMixIn, HTTPServer):
    daemon_threads = True


# Monitoring is enabled by default for all controllers in a namespace and can be disabled by using
# the 'extended-monitoring.deckhouse.io/enabled=false' label
# or the 'extended-monitoring.flant.com/enabled=false' annotation.
def is_monitoring_enabled(labels, annotations):
    if annotations and annotations.get(DEPRECATED_EXTENDED_MONITORING_ENABLED_ANNOTATION, "true") == "false":
        return False
    if labels and labels.get(EXTENDED_MONITORING_ENABLED_LABEL, "true") == "false":
        return False
    return True


def parse_thresholds(labels, annotations, default_thresholds):
    thresholds = copy.deepcopy(default_thresholds)
    if annotations:
        for name, value in annotations.items():
            if name.startswith(DEPRECATED_EXTENDED_MONITORING_ANNOTATION_THRESHOLD_PREFIX):
                unprefixed_name = name.replace(DEPRECATED_EXTENDED_MONITORING_ANNOTATION_THRESHOLD_PREFIX, "")
                thresholds[unprefixed_name] = value
    if labels:
        for name, value in labels.items():
            if name.startswith(EXTENDED_MONITORING_LABEL_THRESHOLD_PREFIX):
                unprefixed_name = name.replace(EXTENDED_MONITORING_LABEL_THRESHOLD_PREFIX, "")
                thresholds[unprefixed_name] = value

    return thresholds


class Annotated(ABC):
    default_thresholds = {}
    # When specified with a watch call, shows changes that occur after that particular version of a resource.
    # Defaults to changes from the beginning of history. When specified for list:
    # - if unset, then the result is returned from remote storage based on quorum-read flag;
    # - if it's 0, then we simply return what we currently have in cache, no guarantee;
    # - if set to non zero, then the result is at least as fresh as given rv.
    default_list_options = {
        "resource_version": 0,
    }

    def __init__(self, namespace, name, kube_labels, kube_annotations):
        self.namespace = namespace
        self.name = name
        self.enabled = is_monitoring_enabled(kube_labels, kube_annotations)
        self.thresholds = parse_thresholds(kube_labels, kube_annotations, self.default_thresholds)

    @classmethod
    def list_threshold_annotated_objects(cls, namespace):
        for kube_object in cls.list(namespace, **cls.default_list_options):
            yield cls(namespace, kube_object.metadata.name, kube_object.metadata.labels,
                      kube_object.metadata.annotations)

    @property
    def formatted(self):
        to_return = ""

        if self.enabled:
            to_return += 'extended_monitoring_{}_enabled{{namespace="{}", {}="{}"}} {}\n'.format(
                self.kind.lower(),
                self.namespace,
                self.kind.lower(),
                self.name, 1)

            if hasattr(self, "thresholds"):
                for k, v in self.thresholds.items():
                    to_return += 'extended_monitoring_{}_threshold{{namespace="{}", threshold="{}", {}="{}"}} {}\n'.format(
                        self.kind.lower(),
                        self.namespace,
                        k,
                        self.kind.lower(),
                        self.name, int(v))
        else:
            to_return += 'extended_monitoring_{}_enabled{{namespace="{}", {}="{}"}} {}\n'.format(
                self.kind.lower(),
                self.namespace,
                self.kind.lower(),
                self.name, 0)

        return to_return

    @property
    @abstractmethod
    def kind(self):
        pass

    @property
    @abstractmethod
    def api(self):
        pass

    @classmethod
    @abstractmethod
    def list(cls, namespace, **kwargs):
        pass


class AnnotatedDeployment(Annotated):
    kind = "Deployment"
    api = kubernetes.client.AppsV1Api()

    @classmethod
    def list(cls, namespace, **kwargs):
        return cls.api.list_namespaced_deployment(namespace, **kwargs).items

    default_thresholds = {
        "replicas-not-ready": 0
    }


class AnnotatedStatefulSet(Annotated):
    kind = "StatefulSet"
    api = kubernetes.client.AppsV1Api()

    @classmethod
    def list(cls, namespace, **kwargs):
        return cls.api.list_namespaced_stateful_set(namespace, **kwargs).items

    default_thresholds = {
        "replicas-not-ready": 0
    }


class AnnotatedDaemonSet(Annotated):
    kind = "DaemonSet"
    api = kubernetes.client.AppsV1Api()

    @classmethod
    def list(cls, namespace, **kwargs):
        return cls.api.list_namespaced_daemon_set(namespace, **kwargs).items

    default_thresholds = {
        "replicas-not-ready": 0
    }


class AnnotatedPod(Annotated):
    kind = "Pod"
    api = kubernetes.client.CoreV1Api()

    @classmethod
    def list(cls, namespace, **kwargs):
        return cls.api.list_namespaced_pod(namespace, **kwargs).items

    default_thresholds = {
        "disk-bytes-warning": 85,
        "disk-bytes-critical": 95,
        "disk-inodes-warning": 85,
        "disk-inodes-critical": 90,
        "container-throttling-warning": 25,
        "container-throttling-critical": 50,
    }


class AnnotatedIngress(Annotated):
    kind = "Ingress"
    api = kubernetes.client.NetworkingV1Api()

    @classmethod
    def list(cls, namespace, **kwargs):
        return cls.api.list_namespaced_ingress(namespace, **kwargs).items

    default_thresholds = {
        "5xx-warning": 10,
        "5xx-critical": 20
    }


class AnnotatedNode(Annotated):
    kind = "Node"
    api = kubernetes.client.CoreV1Api()

    @classmethod
    def list(cls, namespace=None, **kwargs):
        return cls.api.list_node(**kwargs).items

    default_thresholds = {
        "disk-bytes-warning": 70,
        "disk-bytes-critical": 80,
        "disk-inodes-warning": 90,
        "disk-inodes-critical": 95,
        "load-average-per-core-warning": 3,
        "load-average-per-core-critical": 10,
    }


class AnnotatedCronJob(Annotated):
    kind = "CronJob"
    api = kubernetes.client.BatchV1Api()

    @classmethod
    def list(cls, namespace, **kwargs):
        return cls.api.list_namespaced_cron_job(namespace, **kwargs).items


KUBERNETES_OBJECTS = (
    AnnotatedNode,
)
KUBERNETES_NAMESPACED_OBJECTS = (
    AnnotatedDeployment,
    AnnotatedStatefulSet,
    AnnotatedDaemonSet,
    AnnotatedPod,
    AnnotatedIngress,
    AnnotatedCronJob,
)

corev1 = kubernetes.client.CoreV1Api()
apis = kubernetes.client.ApisApi()


def _list_objects(executor, objects, namespace):
    yield from chain.from_iterable(executor.map(lambda k: k.list_threshold_annotated_objects(namespace), objects))


def _get_metrics():
    enabled_nses = []
    quantity = 0

    # iterate over namespaced objects in explicitly enabled via annotation Namespaces
    ns_list = (
        ns.metadata.name for ns in corev1.list_namespace().items
        if
        (ns.metadata.labels and EXTENDED_MONITORING_ENABLED_LABEL in ns.metadata.labels)
        or
        (ns.metadata.annotations and DEPRECATED_EXTENDED_MONITORING_ENABLED_ANNOTATION in ns.metadata.annotations)
    )

    response = """# HELP extended_monitoring_annotations Extended monitoring annotations
      # TYPE extended_monitoring_annotations gauge\n"""

    with ThreadPoolExecutor(max_workers=8) as executor:
        def _list_in_ns(ns):
            enabled_nses.append('\nextended_monitoring_enabled{{namespace="{}"}} 1'.format(ns))
            yield from _list_objects(executor, KUBERNETES_NAMESPACED_OBJECTS, ns)

        for annotated_object in chain.from_iterable(executor.map(_list_in_ns, ns_list)):
            response += annotated_object.formatted
            quantity += 1

        for annotated_object in _list_objects(executor, KUBERNETES_OBJECTS, None):
            response += annotated_object.formatted
            quantity += 1

    response += '\n'.join(enabled_nses)
    quantity += len(enabled_nses)
    return response, quantity


class GetHandler(BaseHTTPRequestHandler):
    _response = ""
    _populated = False
    _last_observe = datetime.now()

    @classmethod
    def get_metrics(cls):
        # setting string variable is atomic in Python
        cls._response, quantity = _get_metrics()
        logging.info('Metrics are collected successfully. Batches quantity: {}'.format(quantity))
        cls._last_observe = datetime.now()

    @classmethod
    def loop_get_metrics(cls):
        cls.get_metrics()
        cls._populated = True
        sleep(30)

        while 1:
            try:
                cls.get_metrics()
            except Exception as loop_err:
                logging.info(str(loop_err))
            sleep(30)

    def do_GET(self):
        if self.path == "/ready":
            apis.get_api_versions()
            # Wait for the first metrics request to succeed
            self.send_response(200 if self.__class__._populated else 400)
            self.end_headers()
            return

        if self.path == "/healthz":
            # Fail if metrics were last collected more than 15 minutes ago
            fresh_enough = self.__class__._last_observe > (datetime.now() - timedelta(minutes=15))
            self.send_response(200 if fresh_enough else 500)
            self.end_headers()
            logging.info('Last observed time: {}'.format(self.__class__._last_observe.strftime("%m/%d/%Y, %H:%M:%S")))
            return

        if self.path == "/metrics":
            self.send_response(200)
            self.send_header('Content-Type',
                             'text/plain; charset=utf-8')
            self.end_headers()
            self.wfile.write(self.__class__._response.encode(encoding="utf-8"))
            return

        self.send_response(404)
        self.end_headers()


if __name__ == '__main__':
    server_address = DEFAULT_SERVER_ADDRESS
    server_port = DEFAULT_PORT

    # Parse host and port
    if len(sys.argv) >= 2:
        server_address = sys.argv[1]
    if len(sys.argv) == 3:
        server_port = int(sys.argv[2])

    server = ThreadingHTTPServer((server_address, server_port), GetHandler)

    try:
        # Run metrics renew in background (daemon thread is canceled on the script exit)
        logging.info('Starting metrics loop')
        Thread(target=GetHandler.loop_get_metrics, daemon=True).start()

        logging.info('Starting server')
        server.serve_forever()
    except Exception as err:
        logging.info('Shutting down server')
        raise err

File: modules/340-extended-monitoring/images/extended-monitoring-exporter/src/extended-monitoring.go
/*
Copyright 2023 Flant JSC

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package extended_monitoring_exporter

import (
	"context"
	"fmt"
	"net/http"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/prometheus/client_golang/api"
	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/prometheus/common/model"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
	metrics "k8s.io/metrics/pkg/client/clientset/versioned"
)

const (
	DeprecatedExtendedMonitoringAnnotationThresholdPrefix = "threshold.extended-monitoring.flant.com/"
	DeprecatedExtendedMonitoringEnabledAnnotation         = "extended-monitoring.flant.com/enabled"
	ExtendedMonitoringLabelThresholdPrefix                = "threshold.extended-monitoring.deckhouse.io/"
	ExtendedMonitoringEnabledLabel                        = "extended-monitoring.deckhouse.io/enabled"
	DefaultServerAddress                                  = "0.0.0.0"
	DefaultPort                                           = 8080
	Pod                                                   = "pod"
	Deployment                                            = "deployment"
	Deploymentstatus                                      = "deploymentstatus"
	Deploymentreplicas                                    = "deploymentreplicas"
	Statefulsetcreated                                    = "statefulsetcreated"
	Statefulsetstatus                                     = "statefulsetstatus"
	Statefulsetreplicas                                   = "statefulsetreplicas"
	Statefulset                                           = "statefulset"
	Daemonset                                             = "daemonset"
	Podcpuusage                                           = "podcpuusage"
	Podmemory                                             = "podmemory"
	Podingress                                            = "podingress"
	Podegress                                             = "podegress"
	Daemonsetcreated                                      = "daemonsetcreated"
	Daemonsetstatus                                       = "daemonsetstatus"
	Deamonsetready                                        = "daemonsetready"
	Daemonsetavailable                                    = "daemonsetavailable"
	Daemonsetunavailable                                  = "daemonsetunavailable"
	Daemonsetunmisscheduled                               = "daemonsetunmisscheduled"
	Nodecpuusage                                          = "nodecpuusage"
	Nodememory                                            = "nodememory"
	Nodestatus                                            = "nodestatus"
	Nodeallocatable                                       = "nodeallocatable"
	Nodecondition                                         = "nodecondition"
	Nodeunschedulable                                     = "nodeunschedulable"
	NodeRole                                              = "node_role"
	NodeLabels                                            = "node_labels"
	Kube_node                                             = "node"
	PrometheusURL                                         = "http://localhost:9090"
)

// These are the metrics that we are going to export to Prometheus
var (
	podMetric = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "extended_monitoring_pod_metric",
			Help: "Pod CPU usage",
		},
		[]string{"pod_name"},
	)

	deployment_metric = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "extended_monitoring_deployment_metric",
			Help: "Deployment status",
		},
		[]string{"deployment_name"},
	)

	deployment_threshold = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "extended_monitoring_deployment_threshold",
			Help: "deployment_threshold",
		},
		[]string{"deployment_name"},
	)

	statefulsetMetric = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "extended_monitoring_statefulset",
			Help: "Statefulset status",
		},
		[]string{"statefulset_name"},
	)
	daemonsetMetric = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "extended_monitoring_daemonset_threshold",
			Help: "Daemonset status",
		},
		[]string{"daemonset_name"},
	)

	podThreshold = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "extended_monitoring_pod_threshold",
			Help: "pod threshold",
		},
		[]string{"pod_name"},
	)

	nodeThreshold = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "extended_monitoring_node_threshold",
			Help: "node_threshold",
		},
		[]string{"statefulset_name"},
	)

	statefulsetThreshold = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "extended_monitoring_statefulset_threshold",
			Help: "statefulset_threshold",
		},
		[]string{"statefulset_name"},
	)

	extended_monitoring_statefulset_enabled = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "extended_monitoring_statefulset_enabled",
			Help: "statefulset_enabled",
		},
		[]string{"statefulset_name"},
	)

	extended_monitoring_node_enabled = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "extended_monitoring_node_enabled",
			Help: "is_node_enabled?",
		},
		[]string{"node_name"},
	)
)

// Register the metrics with Prometheus
func init() {
	prometheus.MustRegister(podMetric)
	prometheus.MustRegister(deployment_metric)
	prometheus.MustRegister(statefulsetMetric)
	prometheus.MustRegister(daemonsetMetric)
	prometheus.MustRegister(podThreshold)
	prometheus.MustRegister(deployment_threshold)
	prometheus.MustRegister(nodeThreshold)
	prometheus.MustRegister(statefulsetThreshold)
	prometheus.MustRegister(extended_monitoring_statefulset_enabled)
	prometheus.MustRegister(extended_monitoring_node_enabled)
}

// exporter is a struct that holds the kubernetes client and the metrics client
type exporter struct {
	client   *kubernetes.Clientset
	metrices *metrics.Clientset
}

type Exporter interface {
	ListAnnotatedObjects(namespace string) ([]Annotated, error)
	GetNamespace() []string
}

func NewExporter() Exporter {
	config, err := rest.InClusterConfig()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create in-cluster config: %v", err)
	}

	clientset, err := kubernetes.NewForConfig(config)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create clientset: %v", err)
	}

	metricsClient, err := metrics.NewForConfig(config)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create metrics client: %v", err)
	}

	return &exporter{
		client:   clientset,
		metrices: metricsClient,
	}
}

/*
This function would connect to the Kubernetes cluster and fetch objects that have been annotated, based on the namespace
and type provided.
*/
func (e *exporter) GetNamespace() []string {
	namespaces, err := e.client.CoreV1().Namespaces().List(context.Background(), metav1.ListOptions{})
	if err != nil {
		fmt.Println("[logs] Error fetching namespaces:", err)
	}

	var ns []string
	for _, namespace := range namespaces.Items {
		ns = append(ns, namespace.Name)
	}
	return ns
}

// ListAnnotatedObjects returns a list of annotated objects in the given namespace that have been annotated with the extended-monitoring.deckhouse.io/enabled=true label.
func (e *exporter) ListAnnotatedObjects(namespace string) ([]Annotated, error) {
	// get the list of objects that have been annotated on the namespaec given with value : extended-monitoring.deckhouse.io/enabled=true
	var annotatedObjects []Annotated
	pods, err := e.client.CoreV1().Pods(namespace).List(context.Background(), metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=true", ExtendedMonitoringEnabledLabel)})
	if err != nil {
		fmt.Println("[logs Pod ] Error: ", err)
		return nil, err
	}
	deployments, err := e.client.AppsV1().Deployments(namespace).List(context.Background(), metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=true", ExtendedMonitoringEnabledLabel)})
	if err != nil {
		fmt.Println("[logs Deployment ] Error: ", err)
		return nil, err
	}
	stateful, err := e.client.AppsV1().StatefulSets(namespace).List(context.Background(), metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=true", ExtendedMonitoringEnabledLabel)})
	if err != nil {
		fmt.Println("[logs Stateful ] Error: ", err)
		return nil, err
	}
	daemons, err := e.client.AppsV1().DaemonSets(namespace).List(context.Background(), metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=true", ExtendedMonitoringEnabledLabel)})
	if err != nil {
		fmt.Println("[logs Daemon ] Error: ", err)
		return nil, err
	}

	nodes, err := e.client.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=true", ExtendedMonitoringEnabledLabel)})
	if err != nil {
		fmt.Println("[logs Node ] Error: ", err)
		return nil, err
	}

	// append all the objects to the annotatedObjects slice
	for _, pod := range pods.Items {

		cputhreshold := pod.Spec.DeepCopy().Overhead.Cpu()
		memorythreshold := pod.Spec.DeepCopy().Overhead.Memory()

		thresholds := make(map[string]string)
		thresholds["CPU"] = cputhreshold.String()
		thresholds["Memory"] = memorythreshold.String()

		annotatedObjects = append(annotatedObjects, Annotated{
			Namespace:  pod.Namespace,
			Name:       pod.Name,
			Object:     "Pod",
			Thresholds: thresholds,
		})

		podMetrics(namespace, pod.Name)
	}

	for _, deployment := range deployments.Items {

		cputhreshold := deployment.Spec.DeepCopy().Template.Spec.Overhead.Cpu()
		memorythreshold := deployment.Spec.DeepCopy().Template.Spec.Overhead.Memory()
		replicas := deployment.Spec.DeepCopy().Replicas

		thresolds := make(map[string]string)

		thresolds["CPU"] = cputhreshold.String()
		thresolds["Memory"] = memorythreshold.String()
		thresolds["Replicas"] = string(*replicas)

		annotatedObjects = append(annotatedObjects, Annotated{
			Namespace:  deployment.Namespace,
			Name:       deployment.Name,
			Object:     "Deployment",
			Thresholds: thresolds,
		})

		deploymentMetrics(namespace, deployment.Name)
	}

	for _, stateful := range stateful.Items {
		cputhreshold := stateful.Spec.DeepCopy().Template.Spec.Overhead.Cpu()
		memorythreshold := stateful.Spec.DeepCopy().Template.Spec.Overhead.Memory()
		replicas := stateful.Spec.DeepCopy().Replicas

		thresolds := make(map[string]string)

		thresolds["CPU"] = cputhreshold.String()
		thresolds["Memory"] = memorythreshold.String()
		thresolds["Replicas"] = string(*replicas)

		annotatedObjects = append(annotatedObjects, Annotated{
			Namespace:  stateful.Namespace,
			Name:       stateful.Name,
			Object:     "StatefulSet",
			Thresholds: thresolds,
		})

		statefulsetMetrics(namespace, stateful.Name)
	}

	for _, daemonset := range daemons.Items {

		cputhreshold := daemonset.Spec.DeepCopy().Template.Spec.Overhead.Cpu()
		memorythreshold := daemonset.Spec.DeepCopy().Template.Spec.Overhead.Memory()

		thresolds := make(map[string]string)

		thresolds["CPU"] = cputhreshold.String()
		thresolds["Memory"] = memorythreshold.String()

		annotatedObjects = append(annotatedObjects, Annotated{
			Namespace: daemonset.Namespace,
			Name:      daemonset.Name,
			Object:    "DaemonSet",
		})

		daemonsetMetrics(namespace, daemonset.Name)
	}

	for _, node := range nodes.Items {
		annotatedObjects = append(annotatedObjects, Annotated{
			Namespace: node.Namespace,
			Name:      node.Name,
			Object:    "Node",
		})

		fmt.Println(node.Name)

		nodeMetrics(node.Name)
	}

	return annotatedObjects, nil

}

type Annotated struct {
	Namespace  string
	Name       string
	Enabled    bool
	Object     string
	Thresholds map[string]string
}

// # Monitoring is enabled by default for all controllers in a namespace and can be disabled by using
// # the 'extended-monitoring.deckhouse.io/enabled=false' label
// # or the 'extended-monitoring.flant.com/enabled=false' annotation.
func (a *Annotated) IsEnabled(kube_labels map[string]string, annotations string) bool {
	if enabled, ok := kube_labels[ExtendedMonitoringEnabledLabel]; ok {
		return enabled == "true"
	} else if enabled, ok := kube_labels[DeprecatedExtendedMonitoringEnabledAnnotation]; ok {
		return enabled == "true"
	} else if enabled, ok := kube_labels[annotations]; ok {
		return enabled == "true"
	}

	return false

}

// ParseThresholds parses the thresholds from the labels and annotations of the kubernetes object.
func (a *Annotated) ParseThresholds(labels map[string]string, annotations map[string]string, defaultThresholds map[string]string) map[string]string {
	thresholds := make(map[string]string)

	for name, value := range annotations {
		if strings.HasPrefix(name, DeprecatedExtendedMonitoringAnnotationThresholdPrefix) {
			prefixedName := strings.Replace(name, DeprecatedExtendedMonitoringAnnotationThresholdPrefix, "", 1)
			thresholds[prefixedName] = value
		}
	}

	for name, value := range labels {
		if strings.HasPrefix(name, ExtendedMonitoringLabelThresholdPrefix) {
			prefixedName := strings.Replace(name, ExtendedMonitoringLabelThresholdPrefix, "", 1)
			thresholds[prefixedName] = value
		}
	}

	return thresholds
}

var SLICE []Annotated
var PODS []Annotated
var DEPLOYMENTS []Annotated
var STATEFULSETS []Annotated
var DAEMONSETS []Annotated
var NODES []Annotated

func main() {
	exporter := NewExporter()
	done := make(chan bool)

	go func() {
		for {
			ns := exporter.GetNamespace()
			for _, namespace := range ns {
				objects, err := exporter.ListAnnotatedObjects(namespace)
				if err != nil {
					fmt.Println(time.Now(), "[logs] Error fetching annotated objects:", err)
				}
				if len(objects) > 0 {
					SLICE = objects
				}
			}
		}
	}()
	var wg sync.WaitGroup

	fmt.Println("Starting server on port http://localhost:8080")

	http.Handle("/metrics", promhttp.Handler())
	http.ListenAndServe(":8080", nil)

	wg.Wait()

	done <- true
}

func podMetrics(namespace, podName string) {

	qeuryCPU := fmt.Sprintf(
		`container_cpu_usage_seconds_total{pod="%s",namespace="%s"}`,
		podName,
		namespace,
	)

	qeuryMemory := fmt.Sprintf(
		`container_memory_usage_bytes{pod="%s",namespace="%s"}`,
		podName,
		namespace,
	)
	queryEgress := fmt.Sprintf(
		`container_network_transmit_bytes_total{pod="%s",interface="eth0"}`,
		podName,
	)

	queryIngress := fmt.Sprintf(
		`container_network_receive_bytes_total{pod="%s",interface="eth0"}`,
		podName,
	)

	go query(qeuryCPU, Podcpuusage, podName, namespace)
	go query(qeuryMemory, Podmemory, podName, namespace)
	go query(queryEgress, Podegress, podName, namespace)
	go query(queryIngress, Podingress, podName, namespace)
}

func deploymentMetrics(namespace, deploymentName string) {
	qeuryStatus := fmt.Sprintf(
		`kube_deployment_status_replicas_available{namespace="%s", deployment="%s"}`,
		namespace,
		deploymentName,
	)

	qeuryReplicas := fmt.Sprintf(
		`kube_deployment_status_replicas{deployment="%s",namespace="%s"}`,
		deploymentName,
		namespace,
	)

	go query(qeuryStatus, Deploymentstatus, deploymentName, namespace)
	go query(qeuryReplicas, Deploymentreplicas, deploymentName, namespace)

}

func statefulsetMetrics(namespace, statefulsetName string) {

	qeuryCreated := fmt.Sprintf(
		`kube_statefulset_created{namespace="%s", statefulset="%s"}`,
		namespace,
		statefulsetName,
	)

	qeuryStatus := fmt.Sprintf(
		`kube_statefulset_status_replicas_available{namespace="%s", statefulset="%s"}`,
		namespace,
		statefulsetName,
	)

	qeuryReplicas := fmt.Sprintf(
		`kube_statefulset_status_replicas{namespace="%s", statefulset="%s"}`,
		namespace,
		statefulsetName,
	)

	go query(qeuryCreated, Statefulsetcreated, statefulsetName, namespace)
	go query(qeuryStatus, Statefulsetstatus, statefulsetName, namespace)
	go query(qeuryReplicas, Statefulsetreplicas, statefulsetName, namespace)

}

func daemonsetMetrics(namespace, daemonsetName string) {
	qeuryCreated := fmt.Sprintf(
		`kube_daemonset_created{namespace="%s", daemonset="%s"}`,
		namespace,
		daemonsetName,
	)

	qeuryStatus := fmt.Sprintf(
		`kube_daemonset_status_number_ready{namespace="%s", daemonset="%s"}`,
		namespace,
		daemonsetName,
	)

	qeuryAvailable := fmt.Sprintf(
		`kube_daemonset_status_number_available{namespace="%s", daemonset="%s"}`,
		namespace,
		daemonsetName,
	)

	qeuryUnmisscheduled := fmt.Sprintf(
		`kube_daemonset_status_number_unavailable{namespace="%s", daemonset="%s"}`,
		namespace,
		daemonsetName,
	)

	qeuryMisscheduled := fmt.Sprintf(
		`kube_daemonset_status_number_misscheduled{namespace="%s", daemonset="%s"}`,
		namespace,
		daemonsetName,
	)

	go query(qeuryCreated, Daemonsetcreated, daemonsetName, namespace)
	go query(qeuryStatus, Daemonsetstatus, daemonsetName, namespace)
	go query(qeuryAvailable, Daemonsetavailable, daemonsetName, namespace)
	go query(qeuryUnmisscheduled, Daemonsetunmisscheduled, daemonsetName, namespace)
	go query(qeuryMisscheduled, Daemonsetunmisscheduled, daemonsetName, namespace)

}

func nodeMetrics(nodeName string) {
	query_nodestatus := fmt.Sprintf(
		`kube_node_status_capacity{unit="core", node="%s"}`,
		nodeName,
	)
	query_node_status_allocatable := fmt.Sprintf(
		`kube_node_status_allocatable{unit="core", node="%s"}`,
		nodeName,
	)

	query_node_status_condition := fmt.Sprintf(
		`kube_node_status_condition{node="%s"}`,
		nodeName,
	)

	query_node_spec_unschedulable := fmt.Sprintf(
		`kube_node_spec_unschedulable{node="%s"}`,
		nodeName,
	)

	query_node_role := fmt.Sprintf(
		`kube_node_role{node="%s"}`,
		nodeName,
	)

	query_node_labels := fmt.Sprintf(
		`kube_node_labels{node="%s"}`,
		nodeName,
	)

	go query(query_nodestatus, Nodestatus, nodeName, "")
	go query(query_node_status_allocatable, Nodeallocatable, nodeName, "")
	go query(query_node_status_condition, Nodecondition, nodeName, "")
	go query(query_node_spec_unschedulable, Nodeunschedulable, nodeName, "")
	go query(query_node_role, NodeRole, nodeName, "")
	go query(query_node_labels, NodeLabels, nodeName, "")
}

// query exports the query to the prometheus server and sets the value of the metric
func query(query string, objectType string, objName string, namespace string) {
	client, err := api.NewClient(api.Config{
		Address: PrometheusURL,
	})
	if err != nil {
		panic(err)
	}

	apiClient := v1.NewAPI(client)

	go func() {
		for {
			result, warning, err := apiClient.QueryRange(context.Background(), query, v1.Range{
				Start: time.Now().Add(-2 * time.Second),
				End:   time.Now(),
				Step:  (2 * time.Second) / 2,
			})

			if err != nil {
				fmt.Println("Error querying Prometheus:", err)
				fmt.Println(objectType)
				continue
			}

			if warning != nil {
				fmt.Println("Warning querying Prometheus:", warning)
			}
			var latestValue float64
			if result != nil && result.Type() == model.ValMatrix {
				matrix := result.(model.Matrix)
				if len(matrix) > 0 {
					latestValue = float64(matrix[matrix.Len()-1].Values[0].Value)
				}
			}
			switch objectType {
			case Podcpuusage:
				podMetric.WithLabelValues(objName + "-cpuusage").Set(latestValue)
			case Podmemory:
				podMetric.WithLabelValues(objName + "-memory").Set(latestValue)
			case Podingress:
				podMetric.WithLabelValues(objName + "-ingress").Set(latestValue)
			case Podegress:
				podMetric.WithLabelValues(objName + "-egress").Set(latestValue)
			case Deploymentstatus:
				deployment_metric.WithLabelValues(objName + "-status").Set(latestValue)
			case Deploymentreplicas:
				deployment_threshold.WithLabelValues(objName + "-replicas").Set(latestValue)
			case Statefulsetcreated:
				statefulsetMetric.WithLabelValues(objName + "-created").Set(latestValue)
			case Statefulsetstatus:
				statefulsetMetric.WithLabelValues(objName + "-status").Set(latestValue)
			case Statefulsetreplicas:
				statefulsetMetric.WithLabelValues(objName + "-replicas").Set(latestValue)
			case Daemonsetcreated:
				daemonsetMetric.WithLabelValues(objName + "-created").Set(latestValue)
			case Daemonsetstatus:
				daemonsetMetric.WithLabelValues(objName + "-status").Set(latestValue)
			case Daemonsetavailable:
				daemonsetMetric.WithLabelValues(objName + "-available").Set(latestValue)
			case Daemonsetunmisscheduled:
				daemonsetMetric.WithLabelValues(objName + "-unmisscheduled").Set(latestValue)
			case Daemonsetunavailable:
				daemonsetMetric.WithLabelValues(objName + "-unavailable").Set(latestValue)
			case Nodestatus:
				nodeThreshold.WithLabelValues(objName + "-status").Set(latestValue)
			case Nodeallocatable:
				nodeThreshold.WithLabelValues(objName + "-allocatable").Set(latestValue)
			case Nodecondition:
				nodeThreshold.WithLabelValues(objName + "-condition").Set(latestValue)
			case Nodeunschedulable:
				nodeThreshold.WithLabelValues(objName + "-unschedulable").Set(latestValue)
			case NodeRole:
				podMetric.WithLabelValues(objName + "-role").Set(latestValue)
			case NodeLabels:
				nodeThreshold.WithLabelValues(objName + "-labels").Set(latestValue)
			default:
				fmt.Println("invalid object type: ", objectType)
			}
		}
	}()
}
